File: hsmem/hscore/storage/resource_layer.py
"""
Resource Layer - 资源层

存储原始多模态数据，是记忆系统的基础层
"""

import json
import uuid
from typing import Dict, Any, Optional
from pathlib import Path
from datetime import datetime


class ResourceLayer:
    """资源层 - 存储原始数据"""

    def __init__(self, base_path: Path):
        self.base_path = base_path
        self.base_path.mkdir(parents=True, exist_ok=True)

    async def store(self, resource_data: Dict[str, Any],
                   modality: str = "text") -> str:
        """
        存储资源

        Args:
            resource_data: 资源数据
            modality: 模态类型

        Returns:
            资源ID
        """
        resource_id = str(uuid.uuid4())

        resource = {
            "id": resource_id,
            "modality": modality,
            "data": resource_data,
            "created_at": datetime.utcnow().isoformat(),
            "metadata": {
                "size": len(json.dumps(resource_data))
            }
        }

        # 按模态类型组织存储
        modality_path = self.base_path / modality
        modality_path.mkdir(exist_ok=True)

        file_path = modality_path / f"{resource_id}.json"
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(resource, f, ensure_ascii=False, indent=2)

        return resource_id

    async def get(self, resource_id: str) -> Optional[Dict[str, Any]]:
        """获取资源"""
        # 在所有模态目录中搜索
        for modality_dir in self.base_path.iterdir():
            if modality_dir.is_dir():
                file_path = modality_dir / f"{resource_id}.json"
                if file_path.exists():
                    with open(file_path, 'r', encoding='utf-8') as f:
                        return json.load(f)
        return None

    def count(self) -> int:
        """统计资源数量"""
        count = 0
        for modality_dir in self.base_path.iterdir():
            if modality_dir.is_dir():
                count += len(list(modality_dir.glob("*.json")))
        return count

File: main/backend/src/main/java/com/heartsphere/memory/service/hsmem/local/store/ResourceLayer.java
package com.heartsphere.memory.service.hsmem.local.store;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * 资源层 - 存储原始多模态数据，与 HSMem Python ResourceLayer 行为一致。
 */
@Component
@RequiredArgsConstructor
@Slf4j
public class ResourceLayer {

    private final ObjectMapper objectMapper;

    /**
     * 存储资源
     *
     * @param basePath    存储根路径（resources 父目录）
     * @param resourceData 资源数据
     * @param modality   模态类型（conversation/text/document）
     * @return 资源ID
     */
    public String store(Path basePath, Map<String, Object> resourceData, String modality) throws Exception {
        String resourceId = UUID.randomUUID().toString();
        Path modalityPath = basePath.resolve(modality);
        Files.createDirectories(modalityPath);

        Map<String, Object> resource = new LinkedHashMap<>();
        resource.put("id", resourceId);
        resource.put("modality", modality);
        resource.put("data", resourceData);
        resource.put("created_at", Instant.now().toString());
        resource.put("metadata", Map.of("size", resourceData.toString().length()));

        Path filePath = modalityPath.resolve(resourceId + ".json");
        objectMapper.writerWithDefaultPrettyPrinter().writeValue(filePath.toFile(), resource);
        return resourceId;
    }

    /**
     * 获取资源
     */
    public Optional<Map<String, Object>> get(Path basePath, String resourceId) {
        try {
            if (!Files.isDirectory(basePath)) return Optional.empty();
            try (Stream<Path> stream = Files.list(basePath)) {
                List<Path> dirs = stream.filter(Files::isDirectory).collect(Collectors.toList());
                for (Path modalityDir : dirs) {
                    Path filePath = modalityDir.resolve(resourceId + ".json");
                    if (Files.exists(filePath)) {
                        @SuppressWarnings("unchecked")
                        Map<String, Object> resource = objectMapper.readValue(filePath.toFile(), Map.class);
                        return Optional.of(resource);
                    }
                }
            }
        } catch (Exception e) {
            log.warn("ResourceLayer.get failed: resourceId={}", resourceId, e);
        }
        return Optional.empty();
    }

    /**
     * 获取所有资源
     */
    public List<Map<String, Object>> getAll(Path basePath) {
        List<Map<String, Object>> results = new ArrayList<>();
        try {
            if (!Files.isDirectory(basePath)) return results;
            try (Stream<Path> stream = Files.list(basePath)) {
                List<Path> dirs = stream.filter(Files::isDirectory).collect(Collectors.toList());
                for (Path modalityDir : dirs) {
                    try (Stream<Path> files = Files.list(modalityDir)) {
                        files.filter(p -> p.getFileName().toString().endsWith(".json"))
                                .forEach(f -> {
                                    try {
                                        @SuppressWarnings("unchecked")
                                        Map<String, Object> resource = objectMapper.readValue(f.toFile(), Map.class);
                                        results.add(resource);
                                    } catch (Exception e) {
                                        log.warn("ResourceLayer.getAll read failed: {}", f, e);
                                    }
                                });
                    }
                }
            }
        } catch (Exception e) {
            log.warn("ResourceLayer.getAll failed", e);
        }
        return results;
    }

    public int count(Path basePath) {
        int[] count = {0};
        try {
            if (!Files.isDirectory(basePath)) return 0;
            try (Stream<Path> stream = Files.list(basePath)) {
                stream.filter(Files::isDirectory).forEach(modalityDir -> {
                    try (Stream<Path> files = Files.list(modalityDir)) {
                        count[0] += (int) files.filter(p -> p.getFileName().toString().endsWith(".json")).count();
                    } catch (Exception e) {
                        log.warn("ResourceLayer.count dir failed: {}", modalityDir, e);
                    }
                });
            }
        } catch (Exception e) {
            log.warn("ResourceLayer.count failed", e);
        }
        return count[0];
    }
}
